api/types/container/network.go

```diff
@@ -0,0 +1,243 @@
+func PortFrom(num uint16, proto PortProto) Port {
+	return Port{num: num, proto: proto}
+}
+
+func (p Port) Num() uint16 {
+	return p.num
+}
+
+func (p Port) Proto() PortProto {
+	return p.proto
+}
+
+// IsValid reports whether the [Port] is an initialized valid port (not the zero value).
+func (p Port) IsValid() bool {
+	return p != zeroPort.Value()
+}
+
+func (p Port) String() string {
+	switch p {
+	case zeroPort.Value():
+		return "invalid port"
+	default:
+		return fmt.Sprintf("%d/%s", p.num, p.proto)
+	}
+}
+
+func (p Port) MarshalText() ([]byte, error) {
+	return []byte(p.String()), nil
+}
+
+func (p *Port) UnmarshalText(text []byte) error {
+	if len(text) == 0 {
+		*p = zeroPort.Value()
+		return nil
+	}
+
+	port, err := ParsePort(string(text))
+	if err != nil {
+		return err
+	}
+
+	*p = port
+	return nil
+}
+
+func (p Port) Range() PortRange {
+	return PortRange{start: p.num, end: p.num, proto: p.proto}
+}
+
+// PortSet is a collection of structs indexed by [Port].
+type PortSet = map[Port]struct{}
+
+// PortBinding represents a binding between a Host IP address and a Host Port.
+type PortBinding struct {
+	// HostIP is the host IP Address
+	HostIP string `json:"HostIp"`
+	// HostPort is the host port number
+	HostPort string `json:"HostPort"`
+}
+
+// PortMap is a collection of [PortBinding] indexed by [Port].
+type PortMap = map[Port][]PortBinding
+
+// PortRange represents a range of port numbers and a protocol in the format "8000-9000/tcp".
+//
+// The zero port range value, i.e. PortRange{}, is invalid; use [ParsePortRange] to create a valid PortRange value.
+type PortRange struct {
+	start uint16
+	end   uint16
+	proto PortProto
+}
+
+// ParsePortRange parses s as a [PortRange].
+func ParsePortRange(s string) (PortRange, error) {
+	if s == "" {
+		return PortRange{}, errors.New("value is empty")
+	}
+
+	portRange, proto, ok := strings.Cut(s, "/")
+	if !ok {
+		proto = "tcp"
+	}
+
+	start, end, ok := strings.Cut(portRange, "-")
+	startVal, err := parsePortNumber(start)
+	if err != nil {
+		return PortRange{}, fmt.Errorf("invalid start port '%s': %w", start, err)
+	}
+
+	if !ok || start == end {
+		return PortRange{start: startVal, end: startVal, proto: PortProto(proto)}, nil
+	}
+
+	endVal, err := parsePortNumber(end)
+	if err != nil {
+		return PortRange{}, fmt.Errorf("invalid end port '%s': %w", end, err)
+	}
+	if endVal < startVal {
+		return PortRange{}, errors.New("invalid port range: " + s)
+	}
+	return PortRange{start: startVal, end: endVal, proto: PortProto(proto)}, nil
+}
+
+// MustParsePortRange calls [ParsePortRange](s) and panics on error.
+func MustParsePortRange(s string) PortRange {
+	pr, err := ParsePortRange(s)
+	if err != nil {
+		panic(err)
+	}
+	return pr
+}
+
+// PortRangeFrom returns a PortRange with the given start and end port numbers and protocol.
+func PortRangeFrom(start, end uint16, proto PortProto) PortRange {
+	return PortRange{start: start, end: end, proto: proto}
+}
```

Comment:
This function can be used to construct ill-formed PortRange values which violate the invariant `pr.Start() <= pr.End()`.
```go
PortRangeFrom(65535, 100, TCP)
```

I'm afraid this constructor will need to validate the parameters and have an error-return.